### PR TITLE
GH-115709: Invalidate executors when a local variable is changed via frame.f_locals

### PR DESCRIPTION
**Pull request by markshannon**
Skipping news as this is fixing a hypothetical bug, rather than a reported bug.

The new test causes a C assertion failure in a debug build of main.

* Issue: gh-115709

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -2424,6 +2424,9 @@ dummy_func(
                 opcode = executor->vm_data.opcode;
                 oparg = (oparg & ~255) | executor->vm_data.oparg;
                 next_instr = this_instr;
+                if (_PyOpcode_Caches[_PyOpcode_Deopt[opcode]]) {
+                    PAUSE_ADAPTIVE_COUNTER(this_instr[1].counter);
+                }
```
Comment (gvanrossum) [inline Python/bytecodes.c:2429]: How does it connect to the purpose of the PR? Or is it just a drive-by bugfix?

File: Objects/frameobject.c [hunk 2]
```diff
@@ -148,8 +148,9 @@ framelocalsproxy_setitem(PyObject *self, PyObject *key, PyObject *value)
     if (PyUnicode_CheckExact(key)) {
         int i = framelocalsproxy_getkeyindex(frame, key, false);
         if (i >= 0) {
-            _PyLocals_Kind kind = _PyLocals_GetKind(co->co_localspluskinds, i);
+            _Py_Executors_InvalidateDependency(PyInterpreterState_Get(), co, 1);
```
Comment (gvanrossum) [inline Objects/frameobject.c:151]: The way I solved the `_Py_TIER2` dependency elsewhere is to put this call itself inside `#ifdef _Py_TIER2`. Then you don't need the dummy version you added to optimizer.c.

File: Python/optimizer.c [hunk 3]
```diff
@@ -1721,4 +1722,14 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
     }
 }
 
+#else
+
+void
+_Py_Executors_InvalidateDependency(PyInterpreterState *interp, void *obj, int is_invalidation)
```
Comment (gvanrossum) [inline Python/optimizer.c]: I'm not super keen on having a dummy version here -- and if we do this we should probably make a few more dummy versions and remove the occasional `#if _Py_TIER2` sprinkled around elsewhere.

**Review by gvanrossum — APPROVED** [review]
While this isn't incorrect I'm not super happy with the inconsistency of sometimes `#ifdef`ing out the call and other times relying on the dummy version existing. But this works, so do what's expedient to make it into beta 1.

File: Python/optimizer.c (hunk 3 above)
Comment (markshannon) [reply Python/optimizer.c]: Would you prefer this in optimizer.h
```C
#else
#define _Py_Executors_InvalidateDependency(A, B, C) ((void)0)
#endif _Py_TIER2
```
and leave optimizer.c alone?

File: Python/bytecodes.c (hunk 1 above)
Comment (markshannon) [reply Python/bytecodes.c:2429]: It's unconnected.
This PR uncovered the bug, so I'll need to fix it before merging this. Might as well do it here.

**Comment by brandtbucher** [comment]
Is this the same failure that we're seeing on GH-118661?

```
Assertion failed: instruction->op.code == ENTER_EXECUTOR, file C:\Users\brandtbucher\cpython\Python\optimizer.c, line 1610
```

File: Python/optimizer.c (hunk 3 above)
Comment (gvanrossum) [reply Python/optimizer.c]: That's better. Though if this is what Brandt asks about maybe we should split it and put that bugfix in beta1 even if the f_locals bugfix goes to beta2.

**Review by gvanrossum — APPROVED** [review]
LGTM. Will merge this ASAP since CI is currently broken and this fixes it.
